Add baremetal IPI platform

stbenjam opened the pull request:
Hi all :wave: 

This adds experimental support for the baremetal IPI platform.  Details about how to use it are in the docs/ directory.  We have a prerequisite that there is an existing Ironic infrastructure, and that can be deployed with [openshift-metal3/dev-scripts](https://github.com/openshift-metal3/dev-scripts).  We are working on moving that to the bootstrap VM, to avoid the dependency on another repository, but we are blocked until RHCOS has an [updated version of libpod](https://bugzilla.redhat.com/show_bug.cgi?id=1724660).  

Other known issues that we are working on are tracked [here](https://github.com/openshift-metal3/kni-installer/issues).

------------

This PR since it's initial opening has been broken up into some smaller PR's in addition to this one.

- [X] [baremetal: vendor terraform-provider-ironic and provide initial terraform configuration](https://github.com/openshift/installer/pull/1958)
- [x] [bootstrap: add platform-specific configuration for baremetal](https://github.com/openshift/installer/pull/1960)

markmc commented on images/baremetal/Dockerfile.ci:
```diff
@@ -0,0 +1,24 @@
+
+
+FROM registry.svc.ci.openshift.org/origin/4.1:base
+COPY --from=builder /go/src/github.com/openshift-metalkube/kni-installer/bin/kni-install /bin/openshift-install
```
needs updating for openshift/installer destination

russellb commented on Gopkg.toml:
```diff
@@ -107,3 +107,12 @@ required = [
+
+[[constraint]]
+  branch = "master"
+  name = "github.com/metal3-io/baremetal-operator"
```
both of these need to be the `github.com/openshift/` versions

stbenjam replied:
Should we do this on the fork, or only on my openshift/installer branch?

russellb replied:
the fork, not just your branch

stbenjam replied:
For CAPB, it is pulling from openshift, just name is metal3-io, is that right?

```
[[constraint]]
  branch = "master"
  name = "github.com/metal3-io/cluster-api-provider-baremetal"
  source = "github.com/openshift/cluster-api-provider-baremetal/pkg/apis"
```

It's how others are done:

```
[[constraint]]
  branch = "master"
  name = "sigs.k8s.io/cluster-api-provider-azure"
  source = "https://github.com/openshift/cluster-api-provider-azure.git"
```

russellb replied:
yes you're right, so it's just baremetal-operator using the wrong source

stbenjam replied:
https://github.com/openshift-metal3/kni-installer/pull/114/files

russellb commented on cmd/openshift-install/create.go:
```diff
@@ -24,6 +24,8 @@ import (
 	"k8s.io/client-go/tools/clientcmd"
 	clientwatch "k8s.io/client-go/tools/watch"
 
+	"github.com/openshift-metalkube/openshift-installer/pkg/asset/installconfig"
```
I imagine these imports all need to be fixed to make CI work against this PR?

stbenjam replied:
Updated, they were fixed everywhere but here

stbenjam commented:
Builds for other platforms will fail until https://github.com/openshift-metal3/kni-installer/pull/115 is in, go-lint is addressed in https://github.com/openshift-metal3/kni-installer/pull/113.

abhinavdahiya commented on data/data/baremetal/masters/main.tf:
```diff
@@ -0,0 +1,53 @@
+# This null resource is used to wait for the Ironic API to become available before trying to create
+# the master hosts.
+resource "null_resource" "api-wait" {
+  provisioner "local-exec" {
```
FYI, this will not be allowed as this runs commands on user's host.

stbenjam replied:
We weren't particularly happy with this, but had trouble finding another path. Do you have a suggestion for a way we can do this? Once Ironic is on the bootstrap node, we have the problem that one provider's entire infrastructure is brought up by another one.  This isn't really well supported by terraform https://github.com/hashicorp/terraform/issues/2430, and mostly people use hacks like this.

Could we do remote-exec on the bootstrap?

cgwalters replied:
My 2c: The installer should also have a container image as part of the payload that is executed on the bootstrap.  That would pave the way to fixing a *lot* of things, including e.g. having masters provisioned by machine API, etc.

abhinavdahiya replied:
>  That would pave the way to fixing a lot of things, including e.g. having masters provisioned by machine API, etc.

how does that tie into your comment?

cgwalters replied:
edit: Removed that bit since it was distracting

stbenjam replied:
> My 2c: The installer should also have a container image as part of the payload that is executed on the bootstrap. Basically it should be as ergonomic to commit code into the installer git to run on the bootstrap as it is to run on the "deploying machine".

Can you clarify what this would do for us? I'm not entirely sure I understand.  The only thing being run on the deploying machine is curl to check if Ironic is available.

We are in the process of moving Ironic as a container onto the bootstrap VM.  This local exec exists for that reason: this resource waits for the Ironic API to be available, and the master resources have a depends_on.  

This prevents terraform from trying to create the masters until Ironic is available on the bootstrap.  This is a hack many people use when one provider is responsible for bringing up the infrastructure used by another.

stbenjam replied:
The PR for the Ironic on bootstrap is https://github.com/openshift-metal3/kni-installer/pull/100, it's still being worked on by @hardys.

> That would pave the way to fixing a lot of things, including e.g. having masters provisioned by machine API, etc.

If masters were provisioned by machine API, then this would avoid a lot of headaches since we wouldn't need terraform for Ironic at all.

hardys replied:
Also note that openshift-metal3/kni-installer#100 doesn't remove the need to wait for ironic to be ready - just starting the container on the boostrap VM isn't enough, we need the API to be up (which doesn't take long), and also for the RHCOS and deploy ramdisk/kernel images to be downloaded (which can take a while).

In my PR I'm using the same curl approach to wait for all-the-things in terraform, so if that's unacceptable we'll need to agree a solution that is.

stbenjam replied:
Another alternative would be to refactor the pkg/asset/cluster code to support the idea of multiple terraform stages if a platform requires it, i.e. 2 distinct `terraform apply`'s where one brings up bootstrap (with Ironic), and the second utilizing Ironic to create the masters.

markmc replied:
Just noting that we've removed this again in openshift-metal3/kni-installer#133 so it'll be gone when this PR is refreshed

abhinavdahiya commented on data/data/bootstrap/files/usr/local/bin/bootkube.sh.template:
```diff
@@ -109,7 +187,7 @@ then
 		/usr/bin/cluster-kube-apiserver-operator render \
 		--manifest-etcd-serving-ca=etcd-ca-bundle.crt \
-		--manifest-etcd-server-urls={{.EtcdCluster}} \
+		--manifest-etcd-server-urls=$(etcd-server-urls $ETCD_MEMBERS) \
```
fix the logic in go that sets the EtcdCluster value. bash function will not be accepted.

markmc replied:
Removed in openshift-metal3/kni-installer#130 so it'll be gone when this PR is refreshed

abhinavdahiya commented on data/data/baremetal/OWNERS:
```diff
@@ -0,0 +1,5 @@
+# See the OWNERS docs: https://git.k8s.io/community/contributors/guide/owners.md
```
We should discuss how to get here. but not right now.

stbenjam replied:
Does this mean you want all OWNERS files removed?

abhinavdahiya replied:
convert approvers to reviewers in the OWNERS files.

abhinavdahiya commented:
Please reduce the bash and python magic, and make sure there is a **lot** of documentation. It's not acceptable to point to docs in a different repo, I would expect that commits have the detailed documentation.

abhinavdahiya commented on data/data/bootstrap/files/usr/local/bin/bootkube.sh.template:
```diff
@@ -96,6 +96,84 @@ then
+	touch mco-bootstrap.done
+fi
+
+echo "Waiting for etcd member names to be available from SRV ..."
```
Why do we care?

russellb replied:
@celebdor @cybertron

cybertron replied:
We needed this to get the etcd members. If they aren't available from DNS yet we can't continue.

abhinavdahiya replied:
How is that it is working fine for now, and bare metal needs this?

abhinavdahiya replied:
>yet we can't continue.

continue to what?

is there lack of resiliency in your scripts, if yes please fix them.

cybertron replied:
As I understand it, the other platforms are not actually looking up the etcd member names from DNS. They're assuming etc-0,1,2. We can't do that because our hostnames are set by site DHCP, which we don't control, and it's problematic to map arbitrary hostnames to indices in a stable way so that they won't change if, for example, a master goes down and that DNS entry disappears.

We can't continue without the DNS entries because the next step in bootstrapping requires us to pass the etcd server urls, which we don't have until they're available from DNS. It's simply a dependency, not a lack of resiliency.

abhinavdahiya replied:
> As I understand it, the other platforms are not actually looking up the etcd member names from DNS. They're assuming etc-0,1,2.

that's not true,
installer creates those dns records (etcd-$idx) and points them to the hostnames, and passes that (etcd-$idx) to the operators.

cybertron replied:
It sounds like the main point is that we need to be passing these names from the installer, not finding them at bootstrap time. That seems doable, so we'll look into it.

russellb replied:
@celebdor Can you describe the alternatives you have in mind?  On the surface, the contents of the SRV record look pretty simple and could be passed in via the installer instead of dynamically discovered:

```
[core@master-0 ~]$ host -t SRV "_etcd-server-ssl._tcp.ostest.test.metalkube.org"
_etcd-server-ssl._tcp.ostest.test.metalkube.org has SRV record 0 10 2380 etcd-2.ostest.test.metalkube.org.
_etcd-server-ssl._tcp.ostest.test.metalkube.org has SRV record 0 10 2380 etcd-1.ostest.test.metalkube.org.
_etcd-server-ssl._tcp.ostest.test.metalkube.org has SRV record 0 10 2380 etcd-0.ostest.test.metalkube.org.
```

These hostnames are not the ones that are provided by the external DHCP server, so assumptions about the DHCP provided hostname don't seem to be relevant here.

O thing that is different with this platform is that the related A records are not created in advance.  For other platforms, they are created in advance before any of this runs.  In the case of the baremetal platform, all of this internal DNS resolution is based on discoery using mDNS, so these `etcd-$N` records will not be resolvable until the masters come up and run their mDNS publisher.

Is the real issue here instead that something needs to not run until these records are resolvable?

cybertron replied:
I guess since https://github.com/openshift-metal3/dev-scripts/commit/e2b12a7386af18dd189d46f09412489fd1601509 it would be possible to hard-code these entries. As long as we're enforcing specific hostnames we should know ahead of time what the etcd entries will look like.

Ideally we _wouldn't_ enforce hostnames this way and would dynamically look them up, but if we're willing to push that additional requirement onto users we can simplify this a lot.

russellb replied:
@cybertron pointed me to https://github.com/openshift-metal3/dev-scripts/blob/master/assets/files/etc/kubernetes/manifests/mdns-publisher.yaml#L62 which explains how the etcd hostnames are derived from the DHCP provided hostname.  If we statically define this as `etcd-$N`, then we're imposing a naming pattern restriction on master hostnames.

Obviously a hostname pattern restriction isn't ideal, but that seems like a reasonable next step here, and then a todo item to remove that as a restriction for baremetal IPI in the future.

That will at least allow us to remove this magic from bootkube.

markmc replied:
Removed in openshift-metal3/kni-installer#130 so it'll be gone when this PR is refreshed

abhinavdahiya commented on pkg/asset/tls/mcscertkey.go:
```diff
@@ -35,9 +36,10 @@ func (a *MCSCertKey) Generate(dependencies asset.Parents) error {
 		ExtKeyUsages: []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth},
 		Validity:     ValidityTenYears,
-		DNSNames:     []string{hostname},
+		DNSNames:     []string{hostname, installConfig.Config.BareMetal.ApiVIP},
```
needs to be switched.

abhinavdahiya commented on pkg/types/baremetal/metadata.go:
```diff
@@ -0,0 +1,8 @@
+type Metadata struct {
+	LibvirtURI string `json:"libvirt_uri"`
+	IronicURI  string `json:"ironic_uri"`
+	ApiVIP     string `json:"api_vip"`
```
`APIVIP`

abhinavdahiya commented on pkg/types/baremetal/platform.go:
```diff
@@ -0,0 +1,59 @@
+package baremetal
+
+type BMC struct {
```
what's BMC

russellb replied:
bare metal controller, a device in a server that allows you to manage it out of band

(obviously a comment would help here)

stbenjam replied:
Comment is coming https://github.com/openshift-metal3/kni-installer/pull/113

abhinavdahiya commented on pkg/types/baremetal/platform.go:
```diff
@@ -0,0 +1,59 @@
+	DefaultMachinePlatform *MachinePool `json:"defaultMachinePlatform,omitempty"`
+
+	// ApiVIP is the VIP to use for internal API communication
+	ApiVIP string `json:"api_vip"`
```
`APIVIP`

abhinavdahiya commented on pkg/types/machinepools.go:
```diff
@@ -55,6 +56,9 @@ type MachinePoolPlatform struct {
 	Azure *azure.MachinePool `json:"azure,omitempty"`
+
+	// BareMetal is the configuration used when installing on bare metal.
+	BareMetal *baremetal.MachinePool `json:"openstack,omitempty"`
```
`json:"baremetal,omitempty"`

russellb commented on cmd/openshift-install/create.go:
```diff
@@ -108,6 +110,18 @@ var (
+						}
+					}
+				}
+
```
This has now been removed and will be gone in the next PR update: https://github.com/openshift-metal3/kni-installer/pull/136

stbenjam commented:
/retest

stbenjam commented:
@abhinavdahiya  Rebased. More docs being worked on in https://github.com/openshift-metal3/kni-installer/pull/139, they'll be included here once that's merged.

Would you mind having another look? 

Regarding commits, how would you like them structured? I've tried to split them up logically a bit here.

stbenjam commented:
/test e2e-aws-upgrade

stbenjam commented:
/test e2e-aws-scaleup-rhel7

russellb commented:
/test e2e-aws-scaleup-rhel7

stbenjam commented:
/test e2e-aws

abhinavdahiya commented on pkg/asset/machines/master.go:
```diff
@@ -6,6 +6,8 @@ import (
 	"path/filepath"
 
 	"github.com/ghodss/yaml"
+	baremetalapi "github.com/metal3-io/cluster-api-provider-baremetal/pkg/apis"
```
this change set doesn't belong to this commit

abhinavdahiya commented on pkg/asset/machines/worker.go:
```diff
@@ -6,6 +6,8 @@ import (
 	"path/filepath"
 
 	"github.com/ghodss/yaml"
+	baremetalapi "github.com/metal3-io/cluster-api-provider-baremetal/pkg/apis"
```
this changeset doesn't belong to this commit

abhinavdahiya commented:
I would expect https://github.com/openshift/installer/pull/1873/commits/272635cb668a8c3742eccfb7a2bbcdd9f777c0ed

to include details on each file that's being added to bootstrap host in the commit message or in the file itself.

abhinavdahiya commented on pkg/asset/ignition/bootstrap/bootstrap.go:
```diff
@@ -132,6 +132,28 @@ func (a *Bootstrap) Generate(dependencies asset.Parents) error {
+	// Check for optional platform specific files/units
+	platform := installConfig.Config.Platform.Name()
+	platformFilePath := fmt.Sprintf("bootstrap/%s/files", platform)
+	directory, err := data.Assets.Open(platformFilePath)
```
this err is not used.

stbenjam replied:
Being fixed in https://github.com/openshift-metal3/kni-installer/pull/143

abhinavdahiya commented on pkg/asset/ignition/bootstrap/bootstrap.go:
```diff
@@ -132,6 +132,28 @@ func (a *Bootstrap) Generate(dependencies asset.Parents) error {
+		}
+	}
+	platformUnitPath := fmt.Sprintf("bootstrap/%s/systemd/units", platform)
+	directory, err = data.Assets.Open(platformUnitPath)
```
the error is not used.

stbenjam replied:
Being fixed in https://github.com/openshift-metal3/kni-installer/pull/143

abhinavdahiya commented on pkg/asset/ignition/machine/node.go:
```diff
@@ -13,6 +13,12 @@ import (
 // served by the machine config server.
 func pointerIgnitionConfig(installConfig *types.InstallConfig, rootCA []byte, role string) *ignition.Config {
+	ignitionHost := fmt.Sprintf("api-int.%s:22623", installConfig.ClusterDomain())
+	if installConfig.Platform.BareMetal != nil {
```
i would expect a switch on platform.Name()

stbenjam replied:
Being fixed in https://github.com/openshift-metal3/kni-installer/pull/143

abhinavdahiya commented on pkg/asset/installconfig/baremetal/baremetal.go:
```diff
@@ -0,0 +1,165 @@
+		return nil, err
+	}
+
+	err = survey.Ask([]*survey.Question{
```
why is a default no appropriate for this.

cybertron replied:
I don't understand what you mean by this. We have a default, although the behavior is a bit complex. In most cases, it is fine to look up these values from DNS. They're mandatory records. However, there are network architectures where the address in DNS may not be the one we want to use internally, so it is necessary to allow users to override the DNS value.

This is further complicated by the fact that when we ask platform-specific questions we do not have the cluster domain yet because on other platforms the cluster domain has a dependency on platform-specific install-config data. In order for us to look these values up from DNS we need the domain, so to break that cycle we do the lookup later while setting platform defaults.

So the default here is a placeholder that means "use the DNS value", but we don't have the DNS value ahead of time so we can't actually populate that. The same logic is used for the ingress vip below.

abhinavdahiya replied:
The goal for terminal prompts is to ask the questions for fields, without which you cannot move forward.

And therefore my question if most users will use or be happy with an a specific IP for APIVIP and IngressVIP so that this prompt is not useful for most.
eg. if the prompt was for localhost IP, 127.0.0.1 is a good default and we don't need to ask the user that question.

and people who wan the customization can still specify by using the install-config.yaml

russellb replied:
I'd be OK dropping TUI support completely, personally.  I really don't expect anyone to ever use it for the `baremetal` platform.  There's too many details that need to be filled in about each bare metal host for a TUI to be a user friendly path.  I think we could just revisit the TUI later.  In this case, we're not looking at something like 5 questions, it's more like 30+.

abhinavdahiya commented on pkg/asset/installconfig/baremetal/baremetal.go:
```diff
@@ -0,0 +1,165 @@
+		return nil, err
+	}
+
+	err = survey.Ask([]*survey.Question{
```
why is a default no appropriate for this.

abhinavdahiya commented on pkg/asset/installconfig/baremetal/baremetal.go:
```diff
@@ -0,0 +1,165 @@
+		return nil, err
+	}
+
+	// Keep prompting for hosts
```
the installer doesn't ask users for number of replicas for control plane and compute.
for most platforms this is defaulted to 3, 3.
what's the default configuration for bare metal?
and I think it would be best if we prompt for that many only.

stbenjam replied:
Tracked in https://github.com/openshift-metal3/kni-installer/issues/101

russellb replied:
I suggested in another thread that we just drop TUI support completely for now ...

abhinavdahiya commented on pkg/asset/installconfig/baremetal/baremetal.go:
```diff
@@ -0,0 +1,165 @@
+
+func apiVIPValidator(ans interface{}) error {
+	if ans.(string) != baremetaldefaults.APIVIP {
+		return validate.IP(ans.(string))
```
this IP probably should be part of the MachineCIDR / ClusterNetwork / Sevice Network

stbenjam replied:
Tracked in https://github.com/openshift-metal3/kni-installer/issues/142

abhinavdahiya commented on pkg/asset/installconfig/baremetal/baremetal.go:
```diff
@@ -0,0 +1,165 @@
+
+func ingressVIPValidator(ans interface{}) error {
+	if ans.(string) != baremetaldefaults.IngressVIP {
+		return validate.IP(ans.(string))
```
https://github.com/openshift/installer/pull/1873/commits/9355096cb9b3a59ee3dd35987f82443fc315054b#r301664853

abhinavdahiya commented on pkg/asset/installconfig/baremetal/baremetal.go:
```diff
@@ -0,0 +1,165 @@
+}
+
+// interfaceValidator validates if the answer provided is a valid network
+// interface.  net.Interfaces in Go does not let us know if it's a bridge, but
```
extra space.

stbenjam replied:
Being fixed in https://github.com/openshift-metal3/kni-installer/pull/143

abhinavdahiya commented on pkg/asset/installconfig/baremetal/host.go:
```diff
@@ -0,0 +1,87 @@
+	err = survey.Ask([]*survey.Question{
+		{
+			Prompt: &survey.Input{
+				Message: "BMC Address",
```
I think it would be best if we use full form for the abbreviation.

abhinavdahiya commented on pkg/asset/installconfig/baremetal/host.go:
```diff
@@ -0,0 +1,87 @@
+		return nil, err
+	}
+
+	err = survey.Ask([]*survey.Question{
```
what's the best guess most people will want..?

russellb replied:
`default` - this probably doesn't need to be in the TUI.

and in other threads i'm suggesting we drop all of the TUI code ...

abhinavdahiya commented on pkg/asset/machines/master.go:
```diff
@@ -177,6 +178,16 @@ func (m *Master) Generate(dependencies asset.Parents) error {
 		}
 		azure.ConfigMasters(machines, clusterID.InfraID)
+	case baremetaltypes.Name:
+		// FIXME: baremetal
```
can we track the fixme in JIRA or github issues rather than code.

hardys replied:
I think this is a rebase mistake and can be removed.

See https://github.com/openshift-metal3/kni-installer/commit/1128a2ab244ff52d84fa81c89633bf5e0c3c5af2

And it was removed in https://github.com/openshift-metal3/kni-installer/commit/5d1f9a8e364e6d2bd0a9d17b90bc86216d1a36ed#diff-66b4febb36af0238bb6c09e72c1c44ec

stbenjam replied:
Sure, I can clean these up.

There's a few like this one https://github.com/openshift/installer/pull/1873/files#diff-b69d29c6306c28bb385391cd0af1b7fcR81 that have descriptive information and a link to a related issue, can we leave those in code? I feel like it's harder to lose track of it when it's inline.

But happy to remove them all and create issues for all of them if you prefer.

abhinavdahiya commented on pkg/asset/rhcos/image.go:
```diff
@@ -67,6 +68,9 @@ func (i *Image) Generate(p asset.Parents) error {
 		//TODO(serbrech): change to right image once available.
 		osimage = "/resourceGroups/rhcos_images/providers/Microsoft.Compute/images/rhcostestimage"
+	case baremetal.Name:
+		// FIXME: baremetal
```
can we track fixme in jira or github issue please rather than in code.

hardys replied:
I think this one is real, we depend on two images, the QEMU one for the bootstrap VM, and currently the OpenStack one for the baremetal hosts.

I have a WIP solution to this ref https://github.com/openshift-metal3/kni-installer/pull/100/commits/81ebb032f2a5c68ecfbd94129b732597d8baa2a7 so perhaps we can discus the best approach to that when we get to proposing that to openshift/installer

abhinavdahiya commented on pkg/types/baremetal/defaults/platform.go:
```diff
@@ -0,0 +1,67 @@
+
+// Defaults for the baremetal platform.
+const (
+	LibvirtURI         = "qemu:///system"
```
the default is different on the `libvirt` platform. why is this different?

stbenjam replied:
`qemu:///system` is the [appropriate way to talk to localhost's system mode daemon](https://libvirt.org/uri.html#URI_qemu).  Libvirt platform appears to assume [libvirt is available in 192.168.122.1, and prefers to use TCP](https://github.com/openshift/installer/pull/1873/files/377d79c71997710f7e22400ff7f6d4fa2f6ed8d8..9355096cb9b3a59ee3dd35987f82443fc315054b#diff-02cb3da236b573e9e866932657f6463a). I do not know why, but I don't think that is a good default for the baremetal platform.  We cannot assume a user is using the libvirt default network, or that they use 192.168.122.0/24.

eparis replied:
libvirt used qemu:///system until we started trying to launch workers via libvirt. Then we just hard coded the 192 address (which mad a lot of people unset/unusable).

We probably 'could' force people to remove all security from their bootstrap node (as we do for libvirt), but I for one strongly prefer this.

abhinavdahiya replied:
i'm fine with the default, but things like ^^ are helpful to include the commit message.

abhinavdahiya commented on pkg/types/baremetal/platform.go:
```diff
@@ -0,0 +1,66 @@
+	DefaultMachinePlatform *MachinePool `json:"defaultMachinePlatform,omitempty"`
+
+	// APIVIP is the VIP to use for internal API communication
+	APIVIP string `json:"api_vip"`
```
`apiVIP` should be the json key.

stbenjam replied:
Being fixed in https://github.com/openshift-metal3/kni-installer/pull/143

abhinavdahiya commented on pkg/types/baremetal/platform.go:
```diff
@@ -0,0 +1,66 @@
+	APIVIP string `json:"api_vip"`
+
+	// IngressVIP is the VIP to use for ingress traffic
+	IngressVIP string `json:"ingress_vip"`
```
`ingressVIP` should be the json key

stbenjam replied:
Being fixed in https://github.com/openshift-metal3/kni-installer/pull/143

abhinavdahiya commented on pkg/types/installconfig.go:
```diff
@@ -34,6 +35,7 @@ var (
 	// hidden-but-supported platform names. This list isn't presented
 	// to the user in the interactive wizard.
 	HiddenPlatformNames = []string{
+		baremetal.Name,
```
it looks like you need libvirt features built, so this should be part of the libvirt tag setup or it's own.

stbenjam replied:
Can you clarify how this relates to the code you're commenting on? Why do build tags have to do with setting hidden platform names?

abhinavdahiya replied:
platforms in hidden platform list don't show up in the TUI.

and my suggestion was rather than dropping baremetal from the TUI.. add it to the list of platforms based on libvirt build tag so that the TUI is accessible to users.

markmc replied:
I think the idea here was to add an installconfig_baremetal.go like the libvirt one, so a binary built with the baremetal tag would have baremetal included in PlatformNames

abhinavdahiya commented on pkg/types/validation/installconfig.go:
```diff
@@ -236,6 +238,11 @@ func validatePlatform(platform *types.Platform, fldPath *field.Path, openStackVa
 		validate(vsphere.Name, platform.VSphere, func(f *field.Path) field.ErrorList { return vspherevalidation.ValidatePlatform(platform.VSphere, f) })
 	}
+	if platform.BareMetal != nil {
+		validate(baremetal.Name, platform.BareMetal, func(f *field.Path) field.ErrorList {
```
can we compact this to one line like the others.

stbenjam replied:
gofmt fails if this is on one line, perhaps because it's longer than the others?

stbenjam commented:
@abhinavdahiya Thanks for the detailed review, we'll begin addressing the comments. 

Could you give us an idea of how far we are from getting this merged?  Our hope is given this is an experimental platform, you might let us defer addressing some issues and fix them in follow-ups. For example, the platform config is not stabilized. Could we defer addressing the comments about the survey in a separate PR later? We expect for 4.2, that users will provide the platform's configuration as YAML.

The majority of the rest of the comments are code style which I'll address now, but I'd like to avoid multiple rounds of small issues given this PR ends up needing rebasing 2-3 times per day.

abhinavdahiya commented on pkg/types/baremetal/metadata.go:
```diff
@@ -0,0 +1,9 @@
+type Metadata struct {
+	LibvirtURI string `json:"libvirt_uri"`
+	IronicURI  string `json:"ironic_uri"`
+	APIVIP     string `json:"api_vip"`
```
why are the VIPs required in metadata??

stbenjam replied:
https://github.com/openshift-metal3/kni-installer/pull/141

abhinavdahiya commented:
> but I'd like to avoid multiple rounds of small issues given this PR ends up needing rebasing 2-3 times per day.

we can merge https://github.com/openshift/installer/pull/1873/commits/693379674cd74ff62349a04447857705422fd786 and https://github.com/openshift/installer/pull/1873/commits/532aa08fcd615bc8136b58186817b1c8e5566dd6 as-is in a separate PR.

we can merge https://github.com/openshift/installer/pull/1873/commits/272635cb668a8c3742eccfb7a2bbcdd9f777c0ed in a separate PR _while i would love for it to have docs on what those magic scripts do, we can take that as a followup_

> We expect for 4.2, that users will provide the platform's configuration as YAML.

then can we do without the terminal prompt codepath in this initial implementation?

hardys commented:
> we can merge [272635c](https://github.com/openshift/installer/commit/272635cb668a8c3742eccfb7a2bbcdd9f777c0ed) in a separate PR _while i would love for it to have docs on what those magic scripts do, we can take that as a followup_

If you're happy with that approach we can raise an issue to track adding better docs on those scripts (which are all related to our dns implementation) - I can take the issue then collaborate with @celebdor who implemented the majority of those scripts.

> then can we do without the terminal prompt codepath in this initial implementation?

I think that would be reasonable - the aim is definitely to enable the TUI, but we don't have to block on it if that means we can iterate faster on getting the initial experimental support landed so we can iterate on the remaining issues via smaller follow-up PRs.

stbenjam commented:
> > but I'd like to avoid multiple rounds of small issues given this PR ends up needing rebasing 2-3 times per day.
> 
> we can merge [6933796](https://github.com/openshift/installer/commit/693379674cd74ff62349a04447857705422fd786) and [532aa08](https://github.com/openshift/installer/commit/532aa08fcd615bc8136b58186817b1c8e5566dd6) as-is in a separate PR.

1 PR or 2? Currently 1, but happy to split apart:

https://github.com/openshift/installer/pull/1958

stbenjam commented:
/retest

stbenjam commented:
@abhinavdahiya I've rebased this PR including all our latest changes from the fork. We have a lot more documentation now in this PR.

I'm still working on sorting out the FIXME's to remove those comments and get them tracked appropriately.

What else do we need to do to get this PR in a mergeable state?

abhinavdahiya commented:
ping @stbenjam

stbenjam commented:
> ping @stbenjam

This will be rebased later today, some patches we needed just landed in the installer fork this morning.

stbenjam commented:
@abhinavdahiya  This is rebased.  The FIXME's in shared code have been removed and are tracked externally, specifically https://github.com/openshift-metal3/kni-installer/issues/160 and https://github.com/openshift-metal3/kni-installer/issues/37.

stbenjam commented:
/retest

abhinavdahiya commented:
/approve

stbenjam commented:
/test e2e-aws

eparis commented on docs/user/metal/install_ipi.md:
```diff
@@ -0,0 +1,225 @@
+repository to handle configuration of Ironic.
+
+The following PR contains the WIP changes for automating Ironic from
+`openshift-install`: https://github.com/openshift-metal3/kni-installer/pull/100
```
does this really belong in the documentation?

russellb replied:
well, it's docs of an experimental platform, and giving a pointer to the status on a WIP item.  I think it makes sense to include ...

eparis commented on docs/user/metal/install_ipi.md:
```diff
@@ -0,0 +1,225 @@
+  The `hardwareProfile` is the field we have available to change that.  This
+  interface is subject to change.  In the meantime, hardware profiles can be
+  found here:
+  https://github.com/metal3-io/baremetal-operator/blob/master/pkg/hardware/profile.go#L48
```
Using `master` and line numbers is a good way to get out of sync. Do you want to pin this to a specific release? Is there a better place to point this which won't get out of date?

russellb replied:
I can link to a doc once this PR lands: https://github.com/metal3-io/baremetal-operator/pull/261

eparis commented on docs/user/metal/install_ipi.md:
```diff
@@ -0,0 +1,225 @@
+metadata:
+  name: ostest
+compute:
+- name: worker
```
This does not set `baremetal: {}`   so what is it?

eparis replied:
Or do we not need to set that on controlplane?

russellb replied:
it's empty right now

eparis commented on docs/user/metal/install_ipi.md:
```diff
@@ -0,0 +1,225 @@
+    baremetal: {}
+platform:
+  baremetal:
+    apiVIP: 192.168.111.5
```
do you also need to specify the dnsVIP and the ingressVIP? Where does that happen?

russellb replied:
Good question.  Those should be documented.  There's code for getting their default values from DNS, so that's why this example worked.  The DNS VIP in particular really needs to be set, as we don't want to require a DNS record for that one.

eparis replied:
how about ironicuri and libvirturi, do we want to show all of the baremetal platform tunables here, or one some of them?

russellb replied:
`ironicuri` will go away shortly once it runs in the bootstrap VM.

`libvirturi` has a sane default, and I'm not sure even needs to be tunable, so I think leaving it out for now is OK

russellb replied:
`dnsVIP` isn't actually supported yet from config ...

russellb replied:
added `ingressVIP` to doc here: https://github.com/openshift-metal3/kni-installer/pull/163

cybertron replied:
Strictly speaking neither apiVIP nor ingressVIP are required. In most circumstances they can be retrieved from DNS. dnsVIP cannot, however, so that will be mandatory. As Russell noted though, that change hasn't merged yet. I will go add it to the docs in that PR.

eparis commented on docs/user/metal/install_ipi.md:
```diff
@@ -0,0 +1,225 @@
+script.
+
+This will be replaced by `machine-api-operator` integration and the following
+PR: https://github.com/openshift/machine-api-operator/pull/302
```
out of date so soon....

eparis commented on docs/user/metal/install_ipi.md:
```diff
@@ -0,0 +1,225 @@
+`baremetal` platform `Machine` actuator
+([cluster-api-provider-baremetal](https://github.com/metal3-io/cluster-api-provider-baremetal)).
+This is currently handled by the
+[08_deploy_bmo.sh](https://github.com/openshift-metal3/dev-scripts/blob/master/08_deploy_bmo.sh)
```
links to `master` aren't stable

russellb replied:
True, though this should go away ASAP, once machine-api-operator integration is done

eparis commented on images/baremetal/Dockerfile.ci:
```diff
@@ -0,0 +1,24 @@
+COPY --from=builder /go/src/github.com/openshift/installer/bin/openshift-install /bin/openshift-install
+
+RUN yum install --setopt=tsflags=nodocs -y \
+    yum update -y && \
```
are we supposed to be running `yum update` or just rely on UBI refreshes for this?

stbenjam replied:
The other platform images do the same (openstack, libvirt) - we've followed their example here

russellb replied:
This is at least consistent with other Dockerfiles that install some rpms, see `images/installer/Dockerfile.upi.ci` and `images/openstack/Dockerfile.ci`

smarterclayton replied:
You never run yum update ever.

eparis commented on pkg/asset/installconfig/platformcredscheck.go:
```diff
@@ -59,7 +60,7 @@ func (a *PlatformCredsCheck) Generate(dependencies asset.Parents) error {
 		opts.Cloud = ic.Config.Platform.OpenStack.Cloud
 		_, err = clientconfig.GetCloudFromYAML(opts)
-	case libvirt.Name, none.Name, vsphere.Name:
+	case baremetal.Name, libvirt.Name, none.Name, vsphere.Name:
```
should we be checking bmc access here?

stbenjam replied:
I think that's a good point, we probably should.  Can we address in a follow-up? I've filed https://github.com/openshift-metal3/kni-installer/issues/162 to track this.

russellb replied:
yeah, that would be a nice enhancement

eparis commented on pkg/tfvars/baremetal/baremetal.go:
```diff
@@ -0,0 +1,107 @@
+
+	for _, host := range platformHosts {
+		// Get hardware profile
+		if host.HardwareProfile == "default" {
```
Is it wrong that I don't like the defaulting here.  Why didn't we set it to hardware.DefaultProfileName back where we set it to "default" ?

russellb replied:
I guess this handles the case where someone explicitly put `default` in install-config

stbenjam replied:
`hardware.DefaultProfileName` is "unknown", which looks weird to me in the install-config.  Like @russellb mentioned on slack, we're still trying to understand the best way to represent hardware profiles so I'd expect this to change.

eparis commented on pkg/tfvars/baremetal/baremetal.go:
```diff
@@ -0,0 +1,107 @@
+		}
+
+		// BMC Driver Info
+		accessDetails, _ := bmc.NewAccessDetails(host.BMC.Address)
```
alarm, alarm, not checking error. why not?

russellb replied:
yolo

stbenjam replied:
https://github.com/openshift-metal3/kni-installer/pull/164

eparis commented on pkg/tfvars/baremetal/baremetal.go:
```diff
@@ -0,0 +1,107 @@
+		if profile.RootDeviceHints.HCTL != "" {
+			rootDevice["hctl"] = profile.RootDeviceHints.HCTL
+		} else {
+			rootDevice["name"] = profile.RootDeviceHints.DeviceName
```
This feels sort of brittle. What if we ever use the other options? vendor,
wwn_vendor_extension, wwn_with_extension, by_path, serial, wwn,
size, rotational, name, hctl, model
Now I have to use it in the metal3 codebase and come out here and rewrite the tfvar translator. Do we think this would make sense as a helper function in metal3?

russellb replied:
in the install_ipi.md file, the whole hardwareProfile thing is highlighted as an area particularly subject to change while this is still experimental

indeed, some helper code might improve this some, but i'm not even sure we got the right abstraction defined yet, but it seems ok while this is experimental

eparis commented on pkg/types/baremetal/defaults/platform.go:
```diff
@@ -0,0 +1,67 @@
+	IronicURI          = "http://localhost:6385/v1"
+	ExternalBridge     = "baremetal"
+	ProvisioningBridge = "provisioning"
+	HardwareProfile    = "default"
```
Ah so here we are setting the hardwareProfile. Why not use DefaultProfileName ?

stbenjam replied:
https://github.com/openshift/installer/pull/1873#discussion_r304056325

russellb replied:
we could, note that we still need other code that translates "default" into DefaultProfileName though.  That's what the other code handles

eparis commented on docs/user/metal/install_ipi.md:
```diff
@@ -0,0 +1,225 @@
+        bootMACAddress: 00:11:07:4e:f6:70
+        hardwareProfile: default
+    - name: openshift-worker-0
+        role: master
```
Is worker-0 a master?

russellb replied:
copy/paste error

eparis replied:
are we going to fix it?

stbenjam replied:
https://github.com/openshift-metal3/kni-installer/pull/166

This PR needs rebasing pretty quick due to conflicts as things are merged, so would prefer to bring this over in a follow-up rather than kick CI from scratch again. I think there might be a few other doc fixes we'll need to bring over.

eparis commented on pkg/types/baremetal/platform.go:
```diff
@@ -0,0 +1,66 @@
+type Host struct {
+	Name            string `json:"name,omitempty"`
+	BMC             BMC    `json:"bmc"`
+	Role            string `json:"role"`
```
I can't decide how I feel about the "Role" being here. We could instead define the hosts under the `compute` and `controlPlane` platform sections.  What would happens when the replica count is not equal to the count of Hosts? Aren't hosts supposed to be generic? Shouldn't I (eventually) be able to use any unprovisioned host as a master or worker? Is the "Role" really an artifact of the host itself? Is it's silicon special?

stbenjam replied:
It could be special, if a user has a number of heterogeneous servers and wants particular ones dedicated to be masters.  I wouldn't want to move it under compute/controlPlane and specifically require that users declare a role, though.

I think once we have workers being deployed by the installer, we'll be rethinking some of this. I'd expect this to become optional, and if unspecified let Ironic pick a random bare metal host through the Allocation mechanism we use.

russellb replied:
Here's the thinking:

All defined hosts get turned into `BareMetalHost` objects.

Assuming we get more than the replica count for masters, the installer has to know how to choose some subset of hosts to provision as masters.  That's what the role was for.

True, we could move these under the controlPlane and compute sections and we wouldn't need the role anymore.  That's something to consider as a follow-up.  I figure all of this is still subject to change while it's experimental.

eparis replied:
Your thought is that in 4.2 this API is experimental, so I should be ok with it?  Wish we had some way to denote the instability of his one part of the installconfig :-(

russellb replied:
well, "accept it as experimental" more than "be ok with it"  :-)

I tried to document in `docs/user/metal/install_ipi.md` how the install-config is subject to change while this is still experimental.  Recommendations welcome on any other and better ways we could make this clear.

eparis replied:
Ohhh, less terrible idea.  A `[]string` in the `compute` and `controlPlane` bare metal platform block that has a list of names. And those names are the hosts here.  so all of the hosts are in 1 place, like this, but which one goes where is in the block where it most makes sense.

eparis replied:
Assuming I get a promise to re-evaluate this before 4.3/GA I'm ok with it how it stands. Though I don't know how it works if things don't match up. I guess that's a QE test case.

russellb replied:
TBH, I don't even know if the code even looks at `role` yet!  I don't think it does.  I'll file a bug on this now to follow up.

stbenjam replied:
I've filed https://github.com/openshift-metal3/kni-installer/issues/165 to keep track of this.

russellb replied:
https://github.com/openshift/installer/issues/2003

eparis commented:
I think I only see 2 major issues.
1. No error check: https://github.com/openshift/installer/pull/1873#discussion_r304046296
2. The existence of Role inside the Host struct. That is a public API and feels really odd to me.  https://github.com/openshift/installer/pull/1873#discussion_r304057442

russellb commented:
> I think I only see 2 major issues.
> 
>     1. No error check: [#1873 (comment)](https://github.com/openshift/installer/pull/1873#discussion_r304046296)

addressed here: https://github.com/openshift-metal3/kni-installer/pull/164

will come in a rebase

>     2. The existence of Role inside the Host struct. That is a public API and feels really odd to me.  [#1873 (comment)](https://github.com/openshift/installer/pull/1873#discussion_r304057442)

This one is going to be harder to address.  Did my explanation of why `role` is there make sense?  and do you buy that since the `baremetal` platform is experimental, this isn't a stable interface yet and we'll work on finalizing it in 4.3?  This isn't the only thing in our install config that will need to change.

eparis commented:
/lgtm

stbenjam commented:
> > I think I only see 2 major issues.
> > ```
> > 1. No error check: [#1873 (comment)](https://github.com/openshift/installer/pull/1873#discussion_r304046296)
> > ```
> 
> addressed here: [openshift-metal3/kni-installer#164](https://github.com/openshift-metal3/kni-installer/pull/164)
> 
> will come in a rebase

Just pulled in this fix here.

stbenjam commented:
@eparis Needs another /lgtm due to rebasing.  Small conflict in Gopkg.toml and  pkg/asset/machines/worker.go

stbenjam commented:
/retest

eparis commented:
/lgtm

russellb commented:
/retest
